doc/conf.py

```diff
@@ -272,6 +272,7 @@
     # unlinkable
     'CoregistrationUI',
     'IntracranialElectrodeLocator',
+    'SourceEstimateViewer',
```

Comment:
Same here -- the name suggests generic-ness.

Maybe `SourceEstimateViewer` -> `SourceTFRViewer` and `view_stc` -> `view_source_tfr` or something.

Reply:
I wrote it so that at a later date, the GUI-related stuff in `mne.viz.Brain` could be refactored over and all the source estimates would go through the GUI code. I think that would be much, much cleaner but if that's not the direction you want to go, I can change it to be more specific and less flexible. Right now, there's even not implemented errors for all the other kinds of source estimates.

Reply:
Let's use more specific naming, otherwise users will be rightfully confused why there is a `view_stc` GUI that does not work with / accept STCs. And I don't think we've decided between routing everything through this new GUI sometime in the future vs refactoring Brain internally to be cleaner/easier to use and sticking with some variant of it instead.

Reply:
Ok, it already supports non-TFR (time-series) volumetric data so I think using TFR is a bit of a misnomer. It just doesn't support surface source estimates yet. What would you name it @larsoner? I'm happy to go with that.

Reply:
> Ok, it already supports non-TFR (time-series) volumetric data so I think using TFR is a bit of a misnomer.

But the focus here is really on TFRs I think. For standard volumetric time (VolSourceEstimate), I think we should continue to promote/use `vol_stc.plot_3d` / `Brain`.

I have not tried the code, but assuming this image is representative of the functionality:

![image](https://user-images.githubusercontent.com/2365790/195669255-fc4cbd61-8b31-4466-884f-f94fbaa0cd0c.png)

To me it looks like an orthoview + 3D plot. So maybe some name incorporating "volume slicer" would make sense if you do want it to be more generic.

Reply:
That's fine with me. The image is representative but if there's not a frequency axis (it's flat) then the bottom plot will just be a line for the time series.

Reply:
So maybe VolumeSlicerGUI or something would be good?

Reply:
Sure, we can change to that but one comment first:

To address the other point, I think `vol_stc.plot_3d` and `Brain` are great and definitely better for some things right now but I think the cost to support surface source estimates is very marginal; it's just removing the slice plots if there are no volume estimates (i.e. surface only and not mixed) and plotting the data on the surface of a 3D brain, maybe with the option to toggle it on and off for mixed source estimates. That's a pretty minor PR relative to this one.

I really liked how `mne-qt-browser` was rolled out where first it was a beta and then it was the default and that seemed to cause much fewer issues with users. From looking around, there's a lot of code, especially this https://github.com/mne-tools/mne-python/blob/main/mne/viz/backends/_qt.py#L1427-L1470 and this https://github.com/mne-tools/mne-python/blob/main/mne/viz/_brain/callback.py#L12-L75 kind of thing that are trivially easy routing it through the a widget but are super complex routing through the renderer.

So, I think, although `plot_3d` and `Brain` work (for the most part and right now; there are some outstanding display issues), in my opinion, it's worth the small amount of overhead to support the redundancy until either the aforementioned code gets simplified and looks a lot more maintainable or the ground-up build in this PR is stable enough that `Brain` can just handle the 3D rendering stuff and not the GUI actions.

I'm not proposing any major projects, I just thought it was worth documenting my thought process for why I made this PR extensible in the way I did, and suggesting that labeling it appropriately so as to be another option to `Brain`/`plot_3d` might be a nice way make the most use of this code.

Reply:
> or the ground-up build in this PR is stable enough that Brain can just handle the 3D rendering stuff and not the GUI actions.

Historically the GUI design components of Brain have actually been the focus of quite a lot of thought and effort -- I think it must total more than a year or two of FTE by now -- from multiple people spanning years (going all the way back to PySurfer about a decade ago). I think it's important to make sure we keep the existing `Brain` GUI layout, interactions, and programmability as much as possible. Brain and its GUI have a lot of features, and each was added because people had compelling use cases for them -- so we want to continue supporting them.

Just a couple of examples of "old lessons" that we'd need to re-learn for the new GUI based on the images I've seen so far:

1. You have some fmin/fmid/fmax sliders. For Brain, we had to deal with stuff like keyboard input handling; min going bigger than mid (and mid bigger than max) properly; dealing with min==mid and mid==max cases; adding "increase all" and "decrease all" functionality that worked well for people; and so forth.
2. You have a time slider. Brain supports stuff like saving movies; "playing" the SourceEsitame (play/pause/reset); dealing with the case of single time points without breaking; etc.
3. All Brain GUI stuff can be controlled from the command line (and be reflected in the GUI properly) -- it's fully programmatically controllable.

Bringing these to parity with Brain is no small task, and this is just for a few examples. And in terms of dealing with bugs, you worked through a hidden `ravel` fortran-order bug that I'm pretty sure I had to work out for Brain previously.

> I think the cost to support surface source estimates is very marginal; it's just removing the slice plots if there are no volume estimates (i.e. surface only and not mixed) and plotting the data on the surface of a 3D brain, maybe with the option to toggle it on and off for mixed source estimates.

I agree adding some basic support for surfaces wouldn't be too much work. But I think that coming to parity with `Brain` would be a lot of work -- it supports multiple views of multiple hemispheres configurable in horizontal and vertical layouts of volume or surface or mixed source estimates, each configurable in terms of how they appear in multiple ways both (often) in the GUI and (always) from the command line / methods (and I'm probably forgetting lots of other stuff).

By contrast, I would say adding the new stuff from this PR -- three orthoview slices, and a TFR plot in place of a time trace, along with interactions between them -- to `Brain` seems like it might be less work. I could be wrong, though!

> until either the aforementioned code gets simplified and looks a lot more maintainable

So I think this is one place we could improve without losing the lessons, use cases, and work from Brain. `Brain`'s user-facing API is at least acceptable -- the `time_viewer` and `show_traces` are not ideal but they get you a GUI like the new one here, *or* a more bare plotting window, so I don't think we need to rethink/deprecate/revamp the user-facing aspects of Brain (though there are always enhancements to be made).

However, `Brain` *is* a bit ugly from a maintenance/internal implementation perspective (as you have pointed out) because the creation of the GUI elements and the creation of the 3D scene are intertwined. They're both routed through `pyvistaqt.BackgroundPlotter`, plus some hacks to re-layout the window and widgets. I assume this is what you mean about it being difficult to maintain and extend.

I propose we try putting `Brain` through some strategic internal refactoring to address this:

1. Use the new, cleaner GUI abstraction that you implemented to create and lay out the GUI elements (porting over all existing interaction logic, conditions, and command<->interactions from Brain). At the end of the day, the `Brain` class and GUI looks and acts the same to the user -- but internally it's easier to understand and extend.
2. In this process, stop using the `BackgroundPlotter` from PyVistaQt and instead use the `QtInteractor`, which is just another Qt widget (plus whatever off-screen renderer is currently used for Notebooks anyway).

(1) makes things easier to extend (adding orthoviews as options for `views`...?), and (2) would probably make it easier to put 3D brain plots in GUIs. I think at the end of the day this QtInteractor-based class would end up being what you're talking about in terms of a "Brain that just displays stuff with no GUI elements" -- it's really a plotting canvas/item/area where we put our 3D stuff. We wouldn't call it Brain (for backward compat since `mne.viz.Brain` should stay the same monolithic structure it is now), maybe `BrainPlotter` or something...

Reply:
Thanks for engaging on this. That plan sounds much better, I wanted to add the GUI to `Brain` in this PR but I thought it would be easier to reimplement a few things than understand the workarounds from routing through the renderer. I think there's some useful refactoring from this PR that could be shared with the idea of making this code and `Brain` better and less complex while decreasing redundancy.